### DISCUSSION TITLE
[AppDir] Server Side Rendering retrieve cookie set in middleware

### DISCUSSION THREAD

**Discussion by tobidemski** [vercel/next.js#49252] [App Router]
Hello,

im trying to set a cookie in the middleware which I would like to retrieve in my server side rendering.

middleware.ts
```js
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";
import { v4 as uuidv4 } from "uuid";

export function middleware(request: NextRequest) {
  let response = NextResponse.next();

  if (request.nextUrl.pathname == "/contact") {
    const csrfToken = uuidv4();

    const cookieOptions = {
      name: "csrfToken",
      value: csrfToken,
      httpOnly: true,
      maxAge: 60 * 30,
      secure: true,
    };

    response.cookies.set(cookieOptions);

    return response;
  }

  return;
}
```

page.tsx

```js
import { headers, cookies } from "next/headers";

export const metadata = {
  title: "Contact us",
};

export default function Home(props: any) {
  const headersStore = headers();
  const cookiesStore = cookies();

/*Contains the current cookie value (from cookie "csrfToken") from middleware */
  const csrfToken = headersStore.get("set-cookie") ?? "";

/* Contains the value from the last request */
  const cookieCsrfToken = cookiesStore.get('csrfToken')?.value;

  return (<>{csrfToken}</>);
}
```

My problem here is, it looks like that headers are returning values from the response and cookies from the current request.
Is this intended because the docs are saying that headers also use the current request.

headers doc: https://nextjs.org/docs/app/api-reference/functions/headers
cookies doc: https://nextjs.org/docs/app/api-reference/functions/cookies

If all works as intended is this the right way to retrieve the current value of the cookie? (Of course I still have to parse the cookie value from "set-cookie" to get the actual value if i read the cookie from the headers)

Im using next.js v13.4.0 with app dir.

**Comment by xiwoo** [comment]
I had the same problem.

According to the Next.js official document guide, I also set cookies in middleware, but it was a problem that SSR page.tsx could not accurately refer to cookies set in response. As a result of continuous confirmation, it was confirmed that the cookies function provided by Next.js refers to the header of the request contained in the response.

So in my case, if I need a newly set cookie reference in SSR page.tsx, I created a new Response object and solved it.

middleware.ts
```typescript

        const {headers} = request;

        const newHeaders = new Headers(headers);
        const newCookie = newHeaders.get('cookie')?.split('; ').map(ck => {
          const [key, value] = ck.split('=');
          return `${key}=${key === 'changeCookieKey' ? accessToken : value}`
        }).join('; ');

        if(newCookie) {
          newHeaders.set("cookie", newCookie);
        }

        res = NextResponse.next({ request: {headers: newHeaders} } );
        res.cookies.set(newAccessTokenCookiesOption);//You have to set it up in response to change it to the client after loading the page.

``` 
page.tsx
```typescript
import { cookies } from "next/headers";

const Page = () => {

    const cookieStore = cookies();
    cookieStore.get("changeCookieKey");// it works!
    return (
        ...
    )
}


``` 

You may not understand it well by turning the translator, but I hope it helps.
